bot/resources/tags/generators-iterators.md

```diff
@@ -0,0 +1,58 @@
+**Generators and Iterators**
+
+What is a generator and an iterator and what is the difference between each other?
+
+**Generator**
+
+A generator yields one item at a time and generates items only when in demand and can be created with a function that may contain a `yield` or sometimes `return` statements, but they're not bound to functions as you can subclass [collections.abc.Generator](https://docs.python.org/3/library/collections.abc.html#collections.abc.Generator). Here we create an infinite generator that can be an iterator.
+```py
+from typing import Generator, Iterator
+
+
+def infinite_stream(start: int) -> Generator[int, None, None] | Iterator[int]:
+    while True:
+        yield start
+        start += 1
+```
+
+A generator can also be an expression(genexpr).
+
+```py
+>>> print(num for num in range(11))
+<generator object <genexpr> at 0x0000000000000000>
+```
+
+How is a generator any different from a list, for example?
+The generator yields one item at a time and generates items only when in demand. Whereas, in a list, Python reserves memory for the whole list.
+
+**Iterator**
+
+An iterator is any object whose class has `__next__` and `__iter__` methods. An iterator can also be created with the built-in function `iter`
+
+```py
+from typing import NoReturn
+
+
+class Foo:
+    def __init__(self, max: int = 0) -> None:
+        self.max = max
+
+    def __iter__(self) -> "Foo":
+        self.n = 0
+        return self
+
+    def __next__(self) -> int | NoReturn:
+        if self.n <= self.max:
+            result = 2**self.n
+            self.n += 1
+            return result
+        else:
+            raise StopIteration
+```
+The `StopIteration` exception must be raised or the iterator is pronounced broken.
+
```

Comment:
```suggestion
\```py
foo = (x * 2 for x in range(5))
bar = sum(x * 2 for x in range(5))
\```

Note that Generators created using expressions are not functions that return Generators, rather, they are already Generator objects. You can only iterate the contents of `foo` once before it is exhausted.
\```py
list(foo) -> [0, 2, 4, 6, 8]
list(foo) -> []
\```

You can also use `*` to unpack Generators as positional arguments for functions, like any other Iterable object.
\```py
print(*foo) -> 0 2 4 6 8
\```
```
Rationale explained in previous suggestion, here it might be a better use of content to explain some examples on generator and expression usage instead of building custom Iterator classes.

> ### Note: Please unescape the `\` marks before each of the code blocks, I can't find any other way to nest codeblocks inside the suggestion codeblock.